tests/callbacks/test_finetuning_callback.py

```diff
@@ -365,3 +365,118 @@ def forward(self, x):
+
+
+class TestCallbacksRestoreCallback(BaseFinetuning):
+
+    def freeze_before_training(self, pl_module):
+        self.freeze(pl_module.layer[:3])
+
+    def finetune_function(self, pl_module, epoch, optimizer, opt_idx):
+        if epoch >= 1:
+            self.unfreeze_and_add_param_group(pl_module.layer[epoch - 1], optimizer)
+
+
+class FinetuningBoringModel(BoringModel):
+
+    def __init__(self):
+        super().__init__()
+        self.layer = nn.Sequential(nn.Linear(32, 32), nn.Linear(32, 32), nn.Linear(32, 32), nn.Linear(32, 2))
+
+    def configure_optimizers(self):
+        parameters = filter(lambda x: x.requires_grad, self.parameters())
+        optimizer = torch.optim.SGD(parameters, lr=0.1)
+        return optimizer
+
+
+def test_callbacks_restore(tmpdir):
+    """
+    Test callbacks restore is called after optimizers have been re-created
+    but before optimizer states reload
+    """
+
+    seed_everything(42)
+
+    chk = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+
+    model = FinetuningBoringModel()
+    callback = TestCallbacksRestoreCallback()
+
+    trainer_kwargs = dict(limit_train_batches=1, default_root_dir=tmpdir, callbacks=[callback, chk], max_epochs=2)
+
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model)
+
+    # only 1 optimizer
+    assert len(callback._internal_optimizer_metadata) == 1
+
+    # only 2 param groups
+    assert len(callback._internal_optimizer_metadata[0]) == 2
+
+    # original parameters
+    assert callback._internal_optimizer_metadata[0][0] == {
+        'lr': 0.1,
+        'momentum': 0,
+        'dampening': 0,
+        'weight_decay': 0,
+        'nesterov': False,
+        'params': ['layer.3.weight', 'layer.3.bias']
+    }
+
+    # new param group
+    assert callback._internal_optimizer_metadata[0][1] == {
+        'lr': 0.01,
+        'momentum': 0,
+        'dampening': 0,
+        'weight_decay': 0,
+        'nesterov': False,
+        'params': ['layer.0.weight', 'layer.0.bias']
+    }
+
+    trainer_kwargs["max_epochs"] = 3
+    trainer_kwargs["resume_from_checkpoint"] = chk.last_model_path
+
+    trainer = Trainer(**trainer_kwargs)
+    trainer.fit(model)
+
+
+def test_callbacks_restore_backbone(tmpdir):
+    """
+    Test callbacks restore is called after optimizers have been re-created
+    but before optimizer states reload
+    """
+
+    class BackboneBoringModel(BoringModel):
+
+        def __init__(self):
+            super().__init__()
+            self.layer = nn.Linear(32, 2)
+            self.backbone = nn.Linear(32, 32)
+
+        def forward(self, x):
+            return self.layer(self.backbone(x))
+
+    cb = ModelCheckpoint(dirpath=tmpdir, save_last=True)
+
+    # Initialize a trainer
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        limit_train_batches=1,
+        max_epochs=3,
+        progress_bar_refresh_rate=0,
+        callbacks=[cb, BackboneFinetuning(unfreeze_backbone_at_epoch=2)]
+    )
+
+    # Train the model ⚡
```

Comment:
hehe, taken straight from the examples?
we should consider cleaning these comments out, as the code is pretty verbose in what it is doing, so these comments are not necessary :)